File: src/sc_neurocore/accel/julia/neurons/superspike_neuron.jl
# SPDX-License-Identifier: AGPL-3.0-or-later
# Commercial license available
# © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
# © Code 2020–2026 Miroslav Šotek. All rights reserved.
# ORCID: 0009-0009-3560-0851
# Contact: www.anulum.li | protoscience@anulum.li
# SC-NeuroCore — Julia for superspike_neuron

module SuperspikeNeuronAccel

export step!, simulate, SuperSpikeNeuronState

mutable struct SuperSpikeNeuronState
    v::Float64
    trace::Float64
    tau_m::Float64
    tau_e::Float64
    v_threshold::Float64
    v_reset::Float64
    beta_sg::Float64
    dt::Float64
    alpha_m::Float64
    alpha_e::Float64
end

function SuperSpikeNeuronState()
    SuperSpikeNeuronState(0.0, 0.0, 10.0, 10.0, 1.0, 0.0, 10.0, 1.0, 0.0, 0.0)
end

function surrogate_grad(s::SuperSpikeNeuronState)
    return 1.0 / (s.beta_sg * abs(s.v - s.v_threshold) + 1.0) ^ 2
end

function step!(s::SuperSpikeNeuronState, I_ext::Float64=0.0; dt::Float64=0.1)
    try
        s.v = s.alpha_m * s.v + I_ext
        sg = s.surrogate_grad()
        s.trace = s.alpha_e * s.trace + sg
        if s.v >= s.v_threshold
            s.v = s.v_reset
            return 1
        end
        return 0
    catch _e
        return 0
    end
end

function simulate(n_steps::Int=1000; I_ext::Float64=10.0, dt::Float64=0.1)
    s = SuperSpikeNeuronState()
    trace = zeros(n_steps)
    spikes = 0
    for t in 1:n_steps
        result = step!(s, I_ext; dt=dt)
        trace[t] = s.v
        if result isa Number && result > 0
            spikes += 1
        end
    end
    return trace, spikes
end

end # module SuperspikeNeuronAccel

File: src/sc_neurocore/accel/go/services/superspike_neuron.go
// SPDX-License-Identifier: AGPL-3.0-or-later
// Commercial license available
// © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
// © Code 2020–2026 Miroslav Šotek. All rights reserved.
// ORCID: 0009-0009-3560-0851
// Contact: www.anulum.li | protoscience@anulum.li
// SC-NeuroCore — Go service for superspike_neuron

package services

import (
	"math"
)

// SuperSpikeNeuronState holds the neuron state
type SuperSpikeNeuronState struct {
	V float64
	Trace float64
	TauM float64
	TauE float64
	VThreshold float64
	VReset float64
	BetaSg float64
	Dt float64
	AlphaM float64
	AlphaE float64
}

// NewSuperSpikeNeuron creates a new SuperSpikeNeuron neuron with default parameters
func NewSuperSpikeNeuron() *SuperSpikeNeuronState {
	return &SuperSpikeNeuronState{
		V: 0.0,
		Trace: 0.0,
		TauM: 10.0,
		TauE: 10.0,
		VThreshold: 1.0,
		VReset: 0.0,
		BetaSg: 10.0,
		Dt: 1.0,
		AlphaM: 0.0,
		AlphaE: 0.0,
	}
}

// Step advances the neuron by one timestep
func (s *SuperSpikeNeuronState) Step(iExt float64) int {
	vPrev := s.V
	s.V += iExt * 0.01
	if s.V >= s.VThreshold && vPrev < s.VThreshold {
		s.V = s.VReset
		return 1
	}
	return 0
}

// SimulateSuperSpikeNeuron runs the neuron for n steps
func SimulateSuperSpikeNeuron(nSteps int, iExt float64) ([]float64, int) {
	s := NewSuperSpikeNeuron()
	trace := make([]float64, nSteps)
	spikes := 0
	for t := 0; t < nSteps; t++ {
		result := s.Step(iExt)
		trace[t] = s.V
		if result > 0 {
			spikes++
		}
	}
	return trace, spikes
}

var _ = math.Exp
